<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [49]:
import pandas as pd
import psycopg2
import requests
import lxml
from io import StringIO

In [51]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [52]:
# текст запроса
query_3_1 = f'''
    SELECT COUNT(*)
    FROM vacancies
'''

In [53]:
pd.read_sql_query(query_3_1, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\4250958588.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_1, connection)


,count
0,49197


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [54]:
query_3_2 = f'''
    SELECT COUNT(*)
    FROM employers
'''

In [55]:
pd.read_sql_query(query_3_2, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\2766852473.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_2, connection)


,count
0,23501


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [56]:
query_3_3 = f'''
    SELECT COUNT(*)
    FROM areas
'''

In [57]:
pd.read_sql_query(query_3_3, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1017433096.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_3, connection)


,count
0,1362


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [58]:
query_3_4 = f'''
    SELECT COUNT(*)
    FROM industries
'''

In [59]:
pd.read_sql_query(query_3_4, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\2182601291.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_4, connection)


,count
0,294


***

В базе содержится 49197 вакансий, 23501 работодателей, 1362 регионов и 294 сферы деятельности

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [60]:
query_4_1 = f'''
    SELECT
        a.name,
        COUNT(*) AS vac_count
    FROM vacancies v
    LEFT JOIN areas a ON v.area_id = a.id
    GROUP BY a.name
    ORDER BY vac_count DESC
'''

In [61]:
pd.read_sql_query(query_4_1, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\2392356777.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_1, connection)


,name,vac_count
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892
...,...,...
764,Тарко-Сале,1
765,Новоаннинский,1
766,Бирск,1
767,Сасово,1


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [62]:
query_4_2 = f'''
    SELECT COUNT(*)
    FROM vacancies
    WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL
'''

In [63]:
pd.read_sql_query(query_4_2, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1239799774.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_2, connection)


,count
0,24073


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [64]:
query_4_3 = f'''
    SELECT
        ROUND(AVG(salary_from), 0),
        ROUND(AVG(salary_to), 0)
    FROM vacancies
'''

In [65]:
pd.read_sql_query(query_4_3, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1149274800.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_3, connection)


,round,round
0,71065.0,110537.0


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [66]:
query_4_4 = f'''
    SELECT
        schedule,
        employment,
        COUNT(*) AS vac_count
    FROM vacancies
    GROUP BY schedule, employment
    ORDER BY vac_count DESC
'''

In [67]:
pd.read_sql_query(query_4_4, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1165210956.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_4, connection)


,schedule,employment,vac_count
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [68]:
query_4_5 = f'''
    SELECT
        experience
    FROM vacancies
    GROUP BY experience
    ORDER BY COUNT(*)
'''

In [69]:
pd.read_sql_query(query_4_5, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\3179429343.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_5, connection)


,experience
0,Более 6 лет
1,Нет опыта
2,От 3 до 6 лет
3,От 1 года до 3 лет


***

Больше всего вакансий в Москве. Зарплату указывают примерно в половине случаев. Зарплатная вилка приблизительно от 71 до 110 тыс.руб. В большинстве вакансий указаны полный день и полная занятость. Самое частое требование опыта от года до трех лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [70]:
query_5_1 = f'''
    SELECT
        ROW_NUMBER() OVER (ORDER BY COUNT(*) DESC) AS NN,
        e.name
    FROM vacancies v
    LEFT JOIN employers e ON v.employer_id = e.id
    GROUP BY e.name
    ORDER BY COUNT(*) DESC
    LIMIT 5
'''

In [71]:
pd.read_sql_query(query_5_1, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\3609881965.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_1, connection)


,nn,name
0,1,Яндекс
1,2,Ростелеком
2,3,Тинькофф
3,4,СБЕР
4,5,Газпром нефть


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [72]:
query_5_2 = f'''
    SELECT
        a.name,
        COUNT(DISTINCT e.id) AS emp_count,
        COUNT(DISTINCT v.id) AS vac_count
    FROM areas a
    LEFT JOIN employers e ON e.area = a.id
    LEFT JOIN vacancies v ON v.area_id = a.id
    GROUP BY a.id
    ORDER BY vac_count ASC, emp_count DESC
'''

In [73]:
pd.read_sql_query(query_5_2, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1604037893.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_2, connection)


,name,emp_count,vac_count
0,Россия,410,0
1,Казахстан,207,0
2,Московская область,75,0
3,Краснодарский край,19,0
4,Ростовская область,18,0
...,...,...,...
1357,Алматы,721,1892
1358,Новосибирск,573,2006
1359,Минск,1115,2112
1360,Санкт-Петербург,2217,2851


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [74]:
query_5_3 = f'''
    SELECT
        e.name,
        COUNT(DISTINCT v.area_id) AS area_count
    FROM employers e
    LEFT JOIN vacancies v ON v.employer_id = e.id
    GROUP BY e.name
    ORDER BY area_count DESC
'''

In [75]:
pd.read_sql_query(query_5_3, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\3225378204.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_3, connection)


,name,area_count
0,Яндекс,181
1,Ростелеком,152
2,Спецремонт,116
3,Поляков Денис Иванович,88
4,ООО ЕФИН,71
...,...,...
23170,СДЕЛКА,0
23171,Alandr Group,0
23172,СДК,0
23173,Lemon Land Lombard,0


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [76]:
query_5_4 = f'''
    SELECT COUNT(*)
    FROM employers e
    LEFT JOIN employers_industries ei ON ei.employer_id = e.id
    WHERE ei.employer_id IS NULL
'''

In [77]:
pd.read_sql_query(query_5_4, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1121459838.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_4, connection)


,count
0,8419


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [78]:
query_5_5 = f'''
    SELECT e.name
    FROM employers e
    LEFT JOIN employers_industries ei ON ei.employer_id = e.id
    GROUP BY e.id
    HAVING COUNT(ei.employer_id) = 4
    ORDER BY e.name
'''

In [79]:
pd.read_sql_query(query_5_5, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\4078395237.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_5, connection)


,name
0,101 Интернет
1,21vek.by
2,2ГИС
3,2К
4,4 пикселя +
...,...
1133,ЮРИОН
1134,ЮТИП Технологии
1135,ЯКласс
1136,ЯрНео


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [80]:
query_5_6 = f'''
    SELECT COUNT(*)
    FROM employers e
    LEFT JOIN employers_industries ei ON ei.employer_id = e.id
    LEFT JOIN industries i ON ei.industry_id = i.id
    WHERE i.name = 'Разработка программного обеспечения'
'''

In [81]:
pd.read_sql_query(query_5_6, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1803119307.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_6, connection)


,count
0,3553


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [82]:
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url, headers={"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"})
tables = pd.read_html(StringIO(response.text), flavor='lxml')
list_city = tables[0]['Город'].to_list()

In [83]:
query_5_7 = f'''
    SELECT
        CASE
            WHEN GROUPING(a.name) = 1 THEN 'Total'
            ELSE a.name
        END AS name,
        COUNT(DISTINCT v.id) AS vac_count
    FROM areas a
    LEFT JOIN vacancies v ON v.area_id = a.id
    LEFT JOIN employers e ON v.employer_id = e.id
    WHERE a.name = ANY(%s) AND e.name = 'Яндекс'
    GROUP BY ROLLUP(a.name)
    ORDER BY GROUPING(a.name), vac_count;
'''

In [84]:
pd.read_sql_query(query_5_7, connection, params=(list_city,))

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\2053972247.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_7, connection, params=(list_city,))


,name,vac_count
0,Омск,21
1,Челябинск,22
2,Красноярск,23
3,Волгоград,24
4,Пермь,25
5,Казань,25
6,Ростов-на-Дону,25
7,Самара,26
8,Уфа,26
9,Краснодар,30


***

На первом месте по количеству вакансий и количеству регионов, в которых представлены вакансии компании находится компания «Яндекс» (имеет десятки вакансий во всех городах-милионниках).

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [85]:
query_6_1 = f'''
    SELECT COUNT(*)
    FROM vacancies
    WHERE name ILIKE '%data%' OR name ILIKE '%данн%'
'''

In [86]:
pd.read_sql_query(query_6_1, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\2465573546.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_1, connection)


,count
0,1771


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [87]:
query_6_2 = f'''
    SELECT COUNT(*)
    FROM vacancies
    WHERE
        ((name ILIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE ANY (ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']))
        AND (name ILIKE '%junior%' OR experience = 'Нет опыта' OR employment = 'Стажировка')
'''

In [88]:
pd.read_sql_query(query_6_2, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\2662785725.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_2, connection)


,count
0,51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [89]:
query_6_3 = f'''
    SELECT COUNT(*)
    FROM vacancies
    WHERE
        ((name ILIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE ANY (ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']))
        AND key_skills ILIKE ANY (ARRAY['%SQL%', '%postgres%'])
'''

In [90]:
pd.read_sql_query(query_6_3, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\3736634856.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_3, connection)


,count
0,229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [91]:
query_6_4 = f'''
    SELECT COUNT(*)
    FROM vacancies
    WHERE
        ((name ILIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE ANY (ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']))
        AND key_skills ILIKE '%python%'
'''

In [92]:
pd.read_sql_query(query_6_4, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\422940813.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_4, connection)


,count
0,357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [93]:
query_6_5 = f'''
    SELECT AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, '\t', '')) + 1)
    FROM vacancies
    WHERE
        ((name LIKE '%ML%' AND name NOT LIKE '%HTML%')
        OR name ILIKE ANY (ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']))
'''

In [94]:
pd.read_sql_query(query_6_5, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1025630572.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_5, connection)


,avg
0,6.406032


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [95]:
query_6_6 = f'''
    SELECT
        experience,
        AVG(
            CASE
                WHEN salary_from IS NULL THEN salary_to
                WHEN salary_to IS NULL THEN salary_from
                ELSE (salary_from + salary_to) / 2
            END
        )
    FROM vacancies
    WHERE
        ((name LIKE '%ML%' AND name NOT LIKE '%HTML%')
        OR name ILIKE ANY (ARRAY['%data scientist%', '%data science%', '%исследователь данных%', '%machine learning%', '%машинн%обучен%']))
        AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
    GROUP BY experience
'''

In [96]:
pd.read_sql_query(query_6_6, connection)

C:\Users\timof\AppData\Local\Temp\ipykernel_11964\1957544559.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_6, connection)


,experience,avg
0,Нет опыта,74642.857143
1,От 1 года до 3 лет,139674.750000
2,От 3 до 6 лет,243114.666667


In [97]:
connection.close()

***

В базе удалось найти 1771 вакансию, имеющую отношение к данным (из них 51 подходят для начинающего дата-сайентиста). В вакансиях указывают около шести ключевых навыков. В вакансиях для DS средняя зарплата меняется в зависимости от опыта от 75 до 243 тыс.руб.

# Общий вывод по проекту

Больше всего вакансий в Москве и крупных городах. В большинстве вакансий указаны полный день и полная занятость. На первом месте по количеству вакансий и количеству регионов, в которых представлены вакансии компании находится компания «Яндекс». Зарплатная вилка приблизительно от 71 до 110 тыс.руб. В вакансиях для DS средняя зарплата от 75 до 243 тыс.руб. (в зависимости от опыта)